File: frontend/src/services/ExpenseService.js
import axios from 'axios';

const API_URL = '/api/expenses';

class ExpenseService {
  getAllExpenses() {
    return axios.get(API_URL);
  }

  getExpenseById(id) {
    return axios.get(`${API_URL}/${id}`);
  }

  createExpense(expense) {
    return axios.post(API_URL, expense);
  }

  updateExpense(id, expense) {
    return axios.put(`${API_URL}/${id}`, expense);
  }

  deleteExpense(id) {
    return axios.delete(`${API_URL}/${id}`);
  }

  getExpensesByCategory(category) {
    return axios.get(`${API_URL}/category/${category}`);
  }

  getExpensesByDateRange(startDate, endDate) {
    return axios.get(`${API_URL}/date-range?startDate=${startDate}&endDate=${endDate}`);
  }

  getExpensesByCategoryAndDateRange(category, startDate, endDate) {
    return axios.get(`${API_URL}/category/${category}/date-range?startDate=${startDate}&endDate=${endDate}`);
  }
}

export default new ExpenseService();

File: backend/src/main/java/com/example/expensetracker/service/ExpenseService.java
package com.example.expensetracker.service;

import com.example.expensetracker.model.Expense;
import com.example.expensetracker.repository.ExpenseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
public class ExpenseService {

    private final ExpenseRepository expenseRepository;

    @Autowired
    public ExpenseService(ExpenseRepository expenseRepository) {
        this.expenseRepository = expenseRepository;
    }

    public List<Expense> getAllExpenses() {
        return expenseRepository.findAll();
    }

    public Expense getExpenseById(Long id) {
        return expenseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Expense not found with id: " + id));
    }

    public Expense createExpense(Expense expense) {
        return expenseRepository.save(expense);
    }

    public Expense updateExpense(Long id, Expense expenseDetails) {
        Expense expense = getExpenseById(id);
        
        expense.setDescription(expenseDetails.getDescription());
        expense.setAmount(expenseDetails.getAmount());
        expense.setDate(expenseDetails.getDate());
        expense.setCategory(expenseDetails.getCategory());
        expense.setNotes(expenseDetails.getNotes());
        
        return expenseRepository.save(expense);
    }

    public void deleteExpense(Long id) {
        Expense expense = getExpenseById(id);
        expenseRepository.delete(expense);
    }

    public List<Expense> getExpensesByCategory(String category) {
        return expenseRepository.findByCategory(category);
    }

    public List<Expense> getExpensesByDateRange(LocalDate startDate, LocalDate endDate) {
        return expenseRepository.findByDateBetween(startDate, endDate);
    }

    public List<Expense> getExpensesByCategoryAndDateRange(String category, LocalDate startDate, LocalDate endDate) {
        return expenseRepository.findByCategoryAndDateBetween(category, startDate, endDate);
    }
}